// app42/App.js
var Page1 = require("page42/Page1");
var View = require("view42");

var Router = require("route42/Route6");

var $ = require("jquery");

// var AppView = require("./AppView");

var App = module.exports = Page1.extend({
	name: "App",
	// View: AppView,
	instantiate: function(){
		this.route = Router();
		this.root = this;
		this.activeNode = this;

		Page1.prototype.instantiate.apply(this, arguments);
	},
	initialize: function(){
		this.init();
		this.render();

		$(function(){
			this.view.prependTo("body");

			setTimeout(function(){
				$(window).resize();
			}, 0);

		}.bind(this));
	},
	set_home: function(homepage){
		homepage.set_route(this.route);
		this.add(homepage);
		this.home = homepage;
	},
	set_activeNode: function(node){
		this.activeNode = node;
	},
	render: function(){
		// this.view = new this.View({
		// 	app: this
		// }).prependTo("body");
	},
	loadPages: function(req){
		var pageKeys = req.keys(),
			routeString, routeParts, route, pages = [], page;

		for (var i = 0; i < pageKeys.length; i++){
			routeString = pageKeys[i].replace("./", "").replace(".page.js", "");
			routeParts = routeString.split("/");
			if (routeParts[routeParts.length - 1] === routeParts[routeParts.length - 2])
				routeParts.pop();

			routeString = routeParts.join("/");

			// require the page
			page = req(pageKeys[i]);

			if (page instanceof Page1){
				page.app = this;
				pages.push(page);
				route = this.route.add(routeString);
				page.set_route(route);
			}
		}

		for (var i = 0; i < pages.length; i++){
			pages[i].getParentPageFromRoute();
		}
	}
});
