// mod42/v5/Mod5.js
var Base = require("base42/v4");
var is = require("util42").is;
var Set = require("set42/v2");
var Extend = require("extend42/v3");

var Logger = require("log42/v2");

var logger = Logger();
/* Already this module has a massive footprint.  */

var Mod = module.exports = Base.extend({
	name:"Mod5",
	
	set: Set().wrapper,

	log: logger.noop,
	set_log: logger.setter,
	
	config: function(){
		this.constructor.track(this);
		this.config_log.apply(this, arguments);
		this.config_name && this.config_name.apply(this, arguments);
	},
	config_log: function(o){
		if (o && is.def(o.log)){ // could be bool false
			this.set_log(o.log); // set_log() gets added by the Logger()
			delete o.log;
		}
	},

	autoInitialize: true,
	instantiate: function(){
		this.assign.apply(this, arguments);
		this.autoInitialize && this.initialize.apply(this, arguments);
	},
	initialize: function(){},
	clone: function(){
		var clone = Object.create(this);
		clone.instantiate.apply(clone, arguments);
		return clone;
	}
}).assign({
	log: logger.noop,
	set: Set().wrapper,
	extend: Extend().wrapper
});
