// extend42/v2/ExtendMfn2.js
var Mfn = require("mfn42");
var track = require("track42");
var logger = require("log42");

var ExtendMfn2 = module.exports = Mfn.extend({
	name: "ExtendMfn2",
	main: function(Base, o){
		var Ext, args = arguments;
		Ext = Base.createConstructor(this.getName(Base, o));
		this.setupConstructor(Ext, Base);
		this.createPrototype(Ext, Base);
		this.instantiatePrototype(Ext, Base);
		this.setupPrototype(Ext, Base, [].slice.call(args, 1));
		this.then && this.then(Ext, Base); // ORP
		return Ext;
	},
	getName: function(Base, o){
		var name;
		if (o && o.name){
			name = o.name;
			delete o.name; // otherwise it will be assigned to prototype..
			return name;
		} else {
			return Base.name + "Ext";
		}
	},
	setupConstructor: function(Ext, Base){
		Ext.assign(Base);
			track(Ext);
			Ext.base = Base;
			if (Base.log){
				console.info("Previous .log re-installation upon .extend is breaking my shit?");
				logger.install(Ext, Base.log.value); // t, f, or undefined for auto
			}
	},
	createPrototype: function(Ext, Base){
		Ext.prototype = Object.create(Base.prototype);
			track(Ext.prototype);
			Ext.prototype.assign({
				constructor: Ext,
				name: Ext.name[0].toLowerCase() + Ext.name.substring(1),
				// base: Base.prototype,
				// proto: Ext.prototype
			});
			// assign now triggers .assignedTo, which is inconvenient in this case
			Ext.prototype.base = Base.prototype;
			Ext.prototype.proto = Ext.prototype; 
	},
	// ORP
	instantiatePrototype: function(Ext, Base){},
	setupPrototype: function(Ext, Base, args){
		Ext.prototype.set.apply(Ext.prototype, args);
		// if (Ext.prototype.protect)
		// 	Ext.prototype.protect(Base.prototype);
	}
});
