// method42/v2/Method2.js
var Mod = require("mod42/v2");
var is = require("util42").is;


var Method = module.exports = Mod.extend({
	name: "Method2",
	set: function(value){
		if (is.fn(value))
			this.fn = value;
		else if (is.bool(value))
			this.log = value;
	},
	initialize: function(){
		this.init_wrapper();
		this.init();
	},
	init_wrapper: function(){
		if (!this.fn){
			this.fn = function(){};
			console.warn("You probably want to wrap something");
		}

		var method = this;

		var wrapper = function(){
			return method.exec(this, arguments); // notice this isn't applied, but just passed along
		};

		wrapper.wrapped = true;
		wrapper.method = this;
		wrapper.fn = this.fn;

		if (this.fn.clone){ // for mfn
			console.warn("Attempting to wrap a mfn..");
			wrapper.clone = function(){
				console.warn("Attempting to clone a Method-wrapped mfn...");
				// I think we can Object.create...
			};
		}

		this.wrapper = wrapper;
		return wrapper;
	},
	exec: function(ctx, args){
		var ret;
		console.group(this.name);
		ret = this.fn.apply(ctx, args);
		console.groupEnd();
		return ret;
	}
}).assign({
	method: function(){
		return this.apply(null, arguments).fn;
	}
});

// otherwise, if you say var method = Method.method, then the window becomes the ctx...
Method.method = Method.method.bind(Method);
