// sandbox/Sandbox.js
var view = require("view42");
var util = require("util42");
var is = util.is;

var $ = require("jquery");

var Sandbox = module.exports = view.extend({
	name: "Sandbox",
	addClass: "flex",
	set: {
		other: function(sandbox, value){
			if (is.fn(value))
				sandbox.set_code(value);
		}
	},
	set_code: function(fn){
		this.code = fn;
	},
	content: function(){
		var sandbox = this;
		this.left = view(function(){
			this.css({
				width: "50%"
			})
			sandbox.textarea = view({
				tag: "textarea",
				content: "yea?",
				css: {
					background: "#333",
					color: "#fff",
					height: "100%",
					width: "100%",
					fontFamily: "courier"
				}
			}).$el.keydown(function(e){
				if (e.which === 9){
					var $self = $(this);
					e.preventDefault();

					var code = $self.val();
					var cursorIndex = $self.prop("selectionStart");
					var beforeCursor = code.slice(0, cursorIndex);
					var afterCursor = code.slice(cursorIndex);

					$self.val(beforeCursor + "   " + afterCursor);

					this.focus();
					this.setSelectionRange(cursorIndex + 3, cursorIndex + 3);
				}
			}).keyup(function(e){
				sandbox.eval($(this).val());
			});
		});

		this.right = view(function(){
			this.css({
				width: "50%"
			})
			sandbox.result = view();
		});
	},
	eval: function(code){
		try {
			eval(code);
		} catch(e){
			this.css("border", "3px solid yellow");
			return false;
		}

		this.css("border", "3px solid black");
		this.result.empty();
		this.result.capture(function(){
			eval(code);
		});
	}
})
