// base42/v4/Base4.js
var createConstructor = require("../v1/createConstructor");
var extend = require("../extend");
var track = require("track42");
var is = require("util42").is;

var Logger = require("log42/v2");

var assign = function(){
	var args = [].slice.call(arguments);
	Object.assign.apply(Object, [this].concat(args));
	return this;
};

var Base = createConstructor("Base1");
track(Base);
track(Base.prototype);

Base.assign = Base.prototype.assign = assign;

Base.prototype.name = "base";


Base.assign({
	extend: extend,
	createConstructor: createConstructor,
	track: track,
	log: Logger()
});

Base.prototype.assign({
	log: Logger(),
	instantiate: function(){
		this.assign.apply(this, arguments);
		this.initialize && this.initialize.apply(this, arguments);
	},
	config: function(){
		this.constructor.track(this);
		this.config_log.apply(this, arguments);
		this.config_name && this.config_name.apply(this, arguments);
	},
	config_log: function(o){
		if (o && is.def(o.log)){ // could be bool false
			this.set_log(o.log); // set_log() gets added by the Logger()
			delete o.log;
		}
	},
	clone: function(){
		var clone = Object.create(this);
		clone.instantiate.apply(clone, arguments);
		return clone;
	}
});

module.exports = Base;
