// path42/PathView.js
var View = require("view42");
require("./Path.less");

var PathView = module.exports = View.BarBox.extend({
	addClass: "path",
	behaviors: function(){},
	bindings: function(){
		var path = this.parent;

		path.root.update.add(function(){
			if (path.isActiveNode())
				this.isActiveNode.show();
			else
				this.isActiveNode.hide();

			if (path.isActive())
				this.isActive.show();
			else
				this.isActive.hide();

			if (path.isActiveParent())
				this.isActiveParent.show();
			else
				this.isActiveParent.hide();

			var activeChild = path.activeChild();
			if (activeChild){
				this.activeChild.ID.text(activeChild.id);
				this.activeChild.show();
			} else {
				this.activeChild.hide();
			}
		}.bind(this));

		this.activator.click(function(){
			this.parent.activate();
		}.bind(this));

		this.myToggle();
	},
	bar: function(){
		var path = this.parent.parent;

		this.parent.isActiveNode = View.Icon("flag").hide();
		this.parent.isActive = View.Icon("asterisk").hide();
		this.parent.isActiveParent = View.Icon("pinterest").hide();
		View(path.name + "{" + path.id + "}");
		this.parent.activeChild = View.Flex(function(){
			this.icon = View.Icon("hashtag");
			this.ID = View();
		}).hide();
		this.parent.activator = View.Icon("bolt");
	},
	box: function(){
		var path = this.parent.parent;

		path.each(function(p){
			p.render();
		});
	}
});

// var PathView = module.exports = View.extend({
// 	name: "PathView",
// 	addClass: "item",
// 	preview: View.x({
// 		addClass: "item",
// 		content: function(){
// 			this.isActiveNode.render();
// 			this.isActive.render();
// 			this.isActiveParent.render();
// 			this.label.render();
// 			this.activeChild.render();
// 			this.activator.render();
// 		},
// 		isActiveNode: View.Icon.x("flag").hide().set({
// 			bindings: function(){
// 				var path = this.parent.parent.parent;
// 				path.isActiveNodeQ.append(function(value){
// 					this.sync(value);
// 				}.bind(this));
// 			},
// 			sync: function(isActiveNode){
// 				if (isActiveNode)
// 					this.show();
// 				else
// 					this.hide();
// 			}
// 		}),
// 		isActive: View.Icon.x("asterisk").hide().set({
// 			bindings: function(){
// 				var path = this.parent.parent.parent;
// 				path.isActiveQ.append(function(value){
// 					this.sync(value);
// 				}.bind(this));
// 			},
// 			sync: function(isActive){
// 				if (isActive)
// 					this.show();
// 				else
// 					this.hide();
// 			}
// 		}),
// 		isActiveParent: View.Icon.x("pinterest").hide().set({
// 			bindings: function(){
// 				var path = this.parent.parent.parent;
// 				path.isActiveParentQ.append(function(value){
// 					this.sync(value);
// 				}.bind(this));
// 			},
// 			sync: function(isActiveParent){
// 				if (isActiveParent)
// 					this.show();
// 				else
// 					this.hide();
// 			}
// 		}),
// 		label: View.x(function(){
// 			var path = this.parent.parent.parent;
// 			this.append(path.name + "{" + path.id + "}")
// 				.addClass("label");
// 		}),
// 		activeChild: View.x({
// 			icon: View.Icon.x("hashtag"),
// 			labl: View.x(),
// 			content: function(){
// 				this.icon.render();
// 				this.labl.render();
// 			},
// 			bindings: function(){
// 				var path = this.parent.parent.parent;
// 				path.activeChildQ.append(function(value){
// 					this.sync(value);
// 				}.bind(this));
// 			},
// 			sync: function(activeChild){
// 				// console.log("activeChild", activeChild);
// 				if (activeChild){
// 					this.labl.text(activeChild.id);
// 					this.show();
// 				} else {
// 					this.hide();
// 				}
// 			}
// 		}).hide(),
// 		activator: View.Icon.x({
// 			addClass: "activator",
// 			type: "bolt",
// 			behaviors: function(){
// 				this.click(function(){
// 					var path = this.parent.parent.parent;
// 					path.activate();
// 				}.bind(this))
// 			}
// 		})
// 	}),
// 	children: View.x(function(){
// 		var path = this.parent.parent;
// 		path.each(function(p){
// 			p.render();
// 		});
// 	}),
// 	content: function(){
// 		this.preview.render();
// 		this.children.render();
// 	}
// });
